src/backend/distributed/planner/recursive_planning.c

```diff
@@ -0,0 +1,606 @@
+
+	return NULL;
+}
+
+
+/*
+ * RecursivelyPlanCTEs plans all CTEs in the query by recursively calling the planner
+ * The resulting plan is added to planningContext->subPlanList and CTE references
+ * are replaced by subqueries that call read_intermediate_result, which reads the
+ * intermediate result of the CTE after it is executed.
+ *
+ * Recursive and modifying CTEs are not yet supported and return an error.
+ */
+static DeferredErrorMessage *
+RecursivelyPlanCTEs(Query *query, RecursivePlanningContext *planningContext)
+{
+	ListCell *cteCell = NULL;
+	CteReferenceWalkerContext context = { -1, NIL };
+
+	if (query->hasModifyingCTE)
+	{
+		/* we could easily support these, but it's a little scary */
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "data-modifying statements are not supported in "
+							 "the WITH clauses of distributed queries",
+							 NULL, NULL);
+	}
+
+	if (query->hasRecursive)
+	{
+		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+							 "recursive CTEs are not supported in distributed "
+							 "queries",
+							 NULL, NULL);
+	}
+
+	/* get all RTE_CTEs that point to CTEs from cteList */
+	CteReferenceListWalker((Node *) query, &context);
+
+	foreach(cteCell, query->cteList)
+	{
+		CommonTableExpr *cte = (CommonTableExpr *) lfirst(cteCell);
+		char *cteName = cte->ctename;
+		Query *subquery = (Query *) cte->ctequery;
+		Query *subPlanQuery = copyObject(subquery);
+		uint64 planId = planningContext->planId;
+		uint32 subPlanId = list_length(planningContext->subPlanList);
+		Query *resultQuery = NULL;
+		DistributedSubPlan *subPlan = NULL;
+		ListCell *rteCell = NULL;
+
+		if (ContainsReferencesToOuterQuery(subquery))
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "CTEs that refer to other subqueries are not "
+								 "supported in multi-shard queries",
+								 NULL, NULL);
+		}
+
+		/* we want to be able to handle queries with only intermediate results */
+		if (!EnableRouterExecution)
+		{
+			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+								 "cannot handle CTEs when the router executor "
+								 "is disabled",
+								 NULL, NULL);
+		}
+
+		/* build a subplan for the CTE */
+		resultQuery = BuildSubPlanResultQuery(subquery, planId, subPlanId);
+
+		if (log_min_messages >= DEBUG1)
+		{
+			StringInfo subPlanString = makeStringInfo();
+			pg_get_query_def(subPlanQuery, subPlanString);
+			elog(DEBUG1, "recursively planning CTE %s: %s", cteName, subPlanString->data);
+		}
+
+		/* replace references to the CTE with a subquery that reads results */
+		foreach(rteCell, context.cteReferenceList)
+		{
+			RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rteCell);
+
+			if (rangeTableEntry->rtekind != RTE_CTE)
+			{
+				/* RTE was already replaced and its ctename is NULL */
+				continue;
+			}
+
+			if (strncmp(rangeTableEntry->ctename, cteName, NAMEDATALEN) == 0)
+			{
+				/* change the RTE_CTE into an RTE_SUBQUERY */
```

Comment:
> Instead of using the existing rangeTableEntry can't we allocate a new one and use it? 

We'd have to insert it into the list in the right place, since indexes matter a lot for RTEs, it might get a little dodgy.

> Also, I wouldn't free anything, some pointer that we don't know might be pointing to here for some unrelated purpose?

makes sense